app/components/Editor.tsx

```diff
@@ -258,23 +261,68 @@ function Editor(props: Props, ref: React.RefObject<SharedEditor> | null) {
     }
   }, [onCreateCommentMark, onDeleteCommentMark, comments.orderedData]);
 
+  const updateFullWidthElems = React.useCallback(() => {
+    // ComponentView updates are delayed until it is rendered by React.
+    // This ensures the callback is executed after React updates the DOM.
+    setTimeout(() => {
+      if (!onFullWidthElemsChange) {
+        return;
+      }
+
+      const view = localRef.current?.view;
+      if (!view) {
+        return;
+      }
+
+      const dom = view.dom;
+
+      const fullWidthTables =
+        dom.querySelectorAll<HTMLElement>(".table-full-width");
+      const fullWidthImages =
+        dom.querySelectorAll<HTMLElement>(".image-full-width");
+
+      const fullWidthElems = Array.from([
+        ...fullWidthTables,
+        ...fullWidthImages,
+      ]);
+
+      const previousElems = previousFullWidthElemsRef.current;
+
+      const newElems = fullWidthElems.filter(
+        (elem) => !previousElems.includes(elem)
+      );
+      const obsoleteElems = previousElems.filter(
+        (elem) => !fullWidthElems.includes(elem)
+      );
+
+      const hasChanged = newElems.length > 0 || obsoleteElems.length > 0;
+
+      if (hasChanged) {
+        previousFullWidthElemsRef.current = fullWidthElems;
+        onFullWidthElemsChange(fullWidthElems);
+      }
+    }, 0);
+  }, [onFullWidthElemsChange]);
+
   const handleChange = React.useCallback(
     (event) => {
       onChange?.(event);
       updateHeadings();
       updateComments();
+      updateFullWidthElems();
```

Comment:
It's tough, because I can see why this is needed but I kind of hate the existing pattern with `updateComments` and `updateHeadings` that exists here so I'm quite remiss to extend it further. There is also the problem that this method is querying the dom instead of the editor data which makes it weirder too, did you try getting the data out of the editor?

Reply:
> did you try getting the data out of the editor
Actually, I did not.
I just found the `nodeDOM` and `domAtPos` methods in prosemirror that allow us to get the DOM node - let me try this.

Reply:
>  I'm quite remiss to extend it further

Makes sense. 
Feels like the editor is doing too many things - setting up the necessary prosemirror objects, rendering auxiliary components / toolbars, handling events. I think it'd be cleaner to extract the editor event handling into a separate dispatcher and plug into the editor.

By having an event dispatcher, it becomes possible to register the interested events for document changes. Example of events would be headings_modified, comment_created, etc.
When the document changes, the event dispatcher would receive the prosemirror node and transactions -> figure out what has happened -> call the necessary event handlers with the transformed data.

An example of the change flow would be
- User adds a heading.
- Editor invokes EventDispatcher with the doc node and transactions.
- EventDispatcher determines that a heading was added and checks if any handler is registered for this event.
- From the doc node, EventHandler builds an internal representation of headings that's understood by the rest of the codebase.
- EventHandler dispatches / calls the registered event handler.

This abstracts the prosemirror -> internal transformation from the rest of the codebase. It also provides a cleaner interface to create / remove event handlers into the editor.
The work involved and the touch points for this implementation would be quite large - probably best to progressively support events in the dispatcher instead of a complete re-write.

I'd love to know your thoughts on this. Or perhaps, suggestions for a different approach.

Reply:
@tommoor Did you get a chance to look at the above query?

Reply:
Do you think it's plausible to reapproach this now with the new document context?

Reply:
Yes - It's in my todo list.
I'll update the PR today 👍